### PR TITLE
DataSourceAttribute Implementation

### PR DESCRIPTION
**Pull request by jayaranigarg**

### CODE REVIEW

**Comment by jayaranigarg** [comment]
@harshjain2 : please have a look....

File: src/TestFramework/MSTest.Core/Attributes/VSTestAttributes.cs [hunk 1]
```diff
@@ -58,21 +58,22 @@ public class TestMethodAttribute : Attribute
...
     /// </summary>
     [AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
+    public class DataTestMethodAttribute : TestMethodAttribute
+    {
+    }
+
+        /// <summary>
+        /// The test initialize attribute.
+        /// </summary>
+        [AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
```
Comment (harshjain2) [inline src/TestFramework/MSTest.Core/Attributes/VSTestAttributes.cs]: nit : formatting.

File: src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs [hunk 2]
```diff
@@ -55,21 +57,18 @@ public class TestMethodInfo : ITestMethod
         /// </summary>
         public bool IsRunnable => string.IsNullOrEmpty(this.NotRunnableReason);
 
-        /// <inheritdoc/>
         public ParameterInfo[] ParameterTypes => this.TestMethod.GetParameters();
```
Comment (harshjain2) [inline src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs:60]: Why removed `/// <inheritdoc/>`?

File: src/Adapter/MSTest.CoreAdapter/Execution/TestMethodRunner.cs [hunk 3]
```diff
@@ -214,36 +214,106 @@ internal UnitTestResult[] RunTestMethod()
...
+                                    this.testContext.SetDataRow(null);
+                                }
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            watch.Stop();
+                            var failedResult = new UTF.TestResult();
+                            failedResult.Outcome = UTF.UnitTestOutcome.Error;
+                            failedResult.TestFailureException = ex;
```
Comment (AbhitejJohn) [inline src/Adapter/MSTest.CoreAdapter/Execution/TestMethodRunner.cs]: this would move inside the foreach.

**Comment by cactuaroid** [comment]
@jayaranigarg, thank you for the works. Now `DataTestMethodAttribute` looks same as `TestMethodAttribute`. Which should I specify with `DataRowAttribute`? No explicit guideline is found on reference.
- [DataTestMethodAttribute](https://docs.microsoft.com/ja-jp/dotnet/api/microsoft.visualstudio.testtools.unittesting.datatestmethodattribute?view=mstest-net-1.2.0)
- [TestMethodAttribute](https://docs.microsoft.com/ja-jp/dotnet/api/microsoft.visualstudio.testtools.unittesting.testmethodattribute?view=mstest-net-1.2.0)
- [DataRowAttribute](https://docs.microsoft.com/ja-jp/dotnet/api/microsoft.visualstudio.testtools.unittesting.datarowattribute?view=mstest-net-1.2.0)

Could you put some explanations on [DataTestMethodAttribute](https://docs.microsoft.com/ja-jp/dotnet/api/microsoft.visualstudio.testtools.unittesting.datatestmethodattribute?view=mstest-net-1.2.0) ?

`DataTestMethodAttribute` has been included since [initial commit](https://github.com/microsoft/testfx/blob/73f7ced91297dfe7a1462300bbf593335b14bc2f/src/TestFramework/vnext/DataTestMethodAttribute.cs). `DataTestMethodAttribute` seems to be integrated into `TestMethodAttribute` in your commit. I think `DataTestMethodAttribute` is now obsolete and just for backward compatibility.

[DataRowAttribute example](https://docs.microsoft.com/ja-jp/dotnet/api/microsoft.visualstudio.testtools.unittesting.datarowattribute?view=mstest-net-1.2.0#examples) uses `TestMethodAttribute`, whereas [this example](https://github.com/dotnet/samples/blob/c1ba91d690939449213e410a0089ca6dc42bc33f/core/getting-started/unit-testing-using-mstest/PrimeService.Tests/PrimeService_IsPrimeShould.cs#L25) uses `DataTestMethodAttribute`.

I found the same confusion here.
https://stackoverflow.com/questions/46904953/datatestmethod-vs-testmethod

**Comment by jayaranigarg** [comment]
@cactuaroid : Can you please open a new issue for this question? I would not encourage you to ask questions on threads/pull requests which have been closed.

**Comment by cactuaroid** [comment]
@jayaranigarg, I'm sorry about that, I have posted #614. Thank you.

**Pull request merged**